test/e2e/serial/tests/network_policy.go

```diff
@@ -0,0 +1,200 @@
+// - Ingress and egress traffic to/from other pods in the cluster is restricted.
+// - Ingress traffic to the metrics endpoints is allowed.
+//
+// Full coverage for inter-pod communication is challenging, so we include basic tests to validate the expected behavior.
+
+var _ = Describe("network policies are applied", Ordered, Label("feature:network_policies"), func() {
+	ctx := context.Background()
+	var namespace string
+	var nropObj *nropv1.NUMAResourcesOperator
+	var nroSchedObj *nropv1.NUMAResourcesScheduler
+	var operatorPod, schedulerPod, rteWorkerPod, prometheusPod *corev1.Pod
+
+	BeforeAll(func() {
+		nropObj = objects.TestNRO()
+		Expect(e2eclient.Client.Get(ctx, client.ObjectKeyFromObject(nropObj), nropObj)).To(Succeed())
+
+		nroSchedObj = objects.TestNROScheduler()
+		Expect(e2eclient.Client.Get(ctx, client.ObjectKeyFromObject(nroSchedObj), nroSchedObj)).To(Succeed())
+
+		Expect(nropObj.Status.NodeGroups).ToNot(BeEmpty())
+		namespace = nropObj.Status.NodeGroups[0].DaemonSet.Namespace
+
+		var err error
+		operatorPod, err = deploy.FindNUMAResourcesOperatorPod(ctx, e2eclient.Client, nropObj)
+		Expect(err).ToNot(HaveOccurred())
+
+		schedulerPod, err = deploy.FindNUMAResourcesSchedulerPod(ctx, e2eclient.Client, nroSchedObj)
+		Expect(err).ToNot(HaveOccurred())
+
+		pods, err := e2eclient.K8sClient.CoreV1().Pods(namespace).List(ctx, metav1.ListOptions{
+			LabelSelector: fmt.Sprintf("name=%s", e2etestenv.RTELabelName),
+		})
+		Expect(err).ToNot(HaveOccurred())
+		Expect(pods.Items).ToNot(BeEmpty())
+		rteWorkerPod = &pods.Items[0]
+
+		prometheusPods, err := e2eclient.K8sClient.CoreV1().Pods("openshift-monitoring").List(ctx, metav1.ListOptions{
+			LabelSelector: labels.SelectorFromSet(map[string]string{
+				"app.kubernetes.io/name": "prometheus",
+			}).String(),
+		})
+		Expect(err).ToNot(HaveOccurred())
+		Expect(prometheusPods.Items).ToNot(BeEmpty())
+		prometheusPod = &prometheusPods.Items[0]
+	})
+
+	type trafficCase struct {
+		FromPod     func() *corev1.Pod
+		ToHost      func() string
+		ToPort      string
+		ShouldAllow bool
+		Description string
+	}
+
+	DescribeTable("traffic behavior",
+		func(tc trafficCase) {
+			Expect(tc.FromPod).ToNot(BeNil(), "source pod should not be nil")
+			klog.InfoS("Running traffic test", "description", tc.Description)
+			reachable := trafficTest(e2eclient.K8sClient, ctx, tc.FromPod(), tc.ToHost(), tc.ToPort)
+			klog.InfoS("reachable", "reachable", reachable)
+
+			if tc.ShouldAllow {
+				Expect(reachable).To(BeTrue(), tc.Description)
+			} else {
+				Expect(reachable).To(BeFalse(), tc.Description)
+			}
+		},
+		// Testing operator and operands egress traffic to API server
+		Entry("operator -> API server", trafficCase{
+			FromPod:     func() *corev1.Pod { return operatorPod },
+			ToHost:      func() string { return "$KUBERNETES_SERVICE_HOST" },
+			ToPort:      "$KUBERNETES_SERVICE_PORT",
```

Comment:
injecting env vars can be risky, in the sense of fragile code. But let's improve later